### ISSUE TITLE
Feature request: Lens like utility for stores

### ISSUE THREAD

**Issue by clinuxrulz** [solidjs/solid#506]
Say our application state holds a collection of states of things in the tabs of the application. It can be handy to have a Lens to focus on a subset of the state in order to provide a store for use by the UI in the tabs state. So that rhe end-user does not need to alway push state changes to top level (under the hood the state still reaches the top, it is just hidden boiler plate).

Make like:
```[getProjState, setProjState] = createLens(path, [getAppState, setAppState]);```

In the example above the created setProjState functions calls setAppState inside with a bit extra prepended to its path.

**Comment by clinuxrulz** [comment]
Actually just the focus on the setter side is more handy to have, because you already get the getter for free from ```<For>``` / ```<Index>```. Maybe Lens is not the best name for it.

**Comment by ryansolid** [comment]
I am on board. I think this is a great idea. I've had this conversation a couple of times. And you are right setters are key since the whole reactive system has the ability to easily pull out the getter side. The implementation is what I've mostly been trying to decide given Solid's internally mutable nature. However it's the path we care about rather than the reference. But the key to not killing performance is not to resolve the path every time you set. Otherwise one could wrap this themselves pretty easily.

Something like this (untested, literally just thinking out loud):
```js
function getPath(path, getState) {
  return () => {
    let state = getState();
    while(segment = path.shift()) {
      state = state[segment]
    }
    return state;
  }
}

function createLensLikeThing(path, [getState, setState]) {
  return [getPath(getState, path), (...args) => setState(...path, ...args)]
}
```
But I feel we can leverage our internals more.

**Comment by clinuxrulz** [comment]
Need to partially resolve a path somehow, then resolve the remainder later.

**Comment by ryansolid** [comment]
Yeah the problem is if someone blows away the path higher up you could be holding a reference to an object that is no longer connected to the tree. So the only thing I can think of is track that part of the path in a memo that and use that to get new getter/setters.. Then it only recalculates when the root changes and otherwise shortcuts the work on each nested set. That will take accessing some of the internals and I feel I can build that in but I need to spend some time playing around with it.

**Comment by uonr** [comment]
Maybe help https://github.com/akheron/optics-ts

**Comment by quaderi** [comment]
I like this idea and given the close similarities between solidjs signals and [klyva](https://github.com/merisbahti/klyva/) atoms (which uses optics-ts and works reasonably well for react, though not widely in usage and limited in documentation & community) I think this is a direction worth pursuing.

I agree 100% though that a rough spot for optics-ts (at least in the context of klyva atoms) that I experienced was the handling of writable optional values/prisms where the path was blown up. But even if anything along the path is not allowed to be undefined/optional for the atom to be writable, it would still be quite useful for many cases. I believe there’s possibly also some challenges around discriminated unions.

**Comment by ryansolid** [comment]
I'm moving any new feature ideas to discussions to better facilitate and keep issues to things that need immediate attention or a have a timeline of attacking.
